### ISSUE TITLE
`restic dump` (semi-)silently drops files from tar archives under certain conditions

### ISSUE THREAD

**Issue by jtru** [restic/restic#4307]
As [reported in the forums a while ago](https://forum.restic.net/t/restic-dump-header-field-too-long/6012), a problem lurks in the tar archive export machinery of `restic dump` that will fail to serialize files into the requested archive when certain file metadata (Resource Forks on MacOS, Extended Ettributes on GNU/Linux, etc.) cannot be contained in [PAX-compatible headers due to size constraints](https://cs.opensource.google/go/go/+/refs/heads/master:src/archive/tar/format.go;l=149;drc=2c70690451f1484607a9172a4c24f78ae832dcb0;bpv=1;bpt=1) inherent to the [format](https://cs.opensource.google/go/go/+/master:src/archive/tar/format.go;l=17).

It is easy to demonstrate this problem on an XFS file system (since it supports very/arbitrarily large xattr data sizes), like so:

```bash
## prepare loopback filesystem containing a file with nasty xattr size
$ truncate --size=100M ./xfstest.img
$ /usr/sbin/mkfs.xfs ./xfstest.img 
$ mkdir /tmp/xfstest_mount
$ sudo mount -o loop ./xfstest.img /tmp/xfstest_mount/
$ sudo chown $UID /tmp/xfstest_mount/.
$ touch /tmp/xfstest_mount/file_without_xattr /tmp/xfstest_mount/file_with_xattr
$ for i in {0..100}; do setfattr -n user.testkey_large$i -v "$(openssl rand -base64 45000 | tr -d '[ +\n]')" /tmp/xfstest_mount/file_with_xattr; done

## create restic repo, create snapshot, dump archive, observe failure
$ openssl rand -base64 48 > /tmp/xattrdemorepo.pass
$ export RESTIC_PASSWORD_FILE=/tmp/xattrdemorepo.pass
$ mkdir /tmp/xattrdemorepo
$ export RESTIC_REPOSITORY=/tmp/xattrdemorepo
$ restic init >/dev/null
$ restic backup --quiet /tmp/xfstest_mount/
$ restic dump latest / > /dev/null
Fatal: cannot dump file: Close: archive/tar: header field too long
```

The error message is of suboptimal utility, since it does not hint at the path of the affected file, and leaves the user guessing what just happened and how to maybe avoid the issue by excluding specific files from the dump operation. Furthermore, as hinted by the "Fatal" prefix of the error message, the very first file that cannot be encoded in the tar archive will cause the `dump` operation to terminate, and subsequent files that were to be dumped **not** to be included in the resulting archive.

Performing the same `dump` operation with `-a zip` (to dump a zip archive instead of a tar archive) succeeds - probably because zip does not even try to contain the metadata the tar writer tries to stuff into PAX headers, and without any hints at potential problems.

I find it surprisingly hard to nail down the least worst option to deal with this problem ;) I created a crude patched build of restic (patch included below) that would dump out "broken" (i.e., too long) PAX/tar header structs and log the path of the node/file that caused the `dump` operation to break, and I think the latter would be the absolute minimum required to implement for restic proper.

What I would like to discuss in more detail in this issue is what to do while generating the tar archive in the face of this problem - there's at least the following options that present themselves:

1. continue behaving as before - that is, truncate the dumped tar archive, and exit with failure
2. skip all affected files entirely - i. e., have neither its name, nor contained data, nor associated metadata - and exclude them from the dumped tar archive (but log to stderr that this happened)
3. dump all affected files into the resulting tar archive *without* the metadata/header fields that prove too long to be properly encoded by `archive/tar` to be PAX-compatible
4. truncate all resulting PAX header fields to a suitable size for them do be encoded by `archive/tar`

What do you think would be the best approach?

I am of the opinion that the user should be able to control this behavior (i.e., have the choice between all options by means of a CLI switch), and that `restic dump` should only provide the mechanism to do so, while applying a sensible default policy (the safest would probably be option *3*?).

FTR, the patch I came up with to aid me in debugging this issue - which I happened to encounter in a forensic backup of a Apple file system where a number of directories contained Resource Forks close to 3Mbytes in size each - was this (I am not at all familiar with golang, as you're about to see ;)):

```
--- a/internal/dump/tar.go
+++ b/internal/dump/tar.go
@@ -4,9 +4,12 @@ import (
        "archive/tar"
        "context"
        "os"
+       "encoding/gob"
+       "bytes"
        "path/filepath"
        "strings"
 
+       "github.com/restic/restic/internal/debug"
        "github.com/restic/restic/internal/errors"
        "github.com/restic/restic/internal/restic"
 )
@@ -94,7 +97,16 @@ func (d *Dumper) dumpNodeTar(ctx context.Context, node *restic.Node, w *tar.Writ
 
        err = w.WriteHeader(header)
        if err != nil {
-               return errors.Wrap(err, "TarHeader")
+               var corr_header bytes.Buffer
+               var ofn = "/tmp/broken_tar_header__" + strings.Replace(header.Name, "/", "_", -1)
+               enc := gob.NewEncoder(&corr_header)
+               enc.Encode(header)
+               fo, _ := os.Create(ofn)
+               fo.Write(corr_header.Bytes())
+               fo.Close()
+               debug.Log(">>>>> dumped headers to %v", ofn)
+               debug.Log("ERROR while writing tar headers for file: %v", node.Path)
+               err = nil
        }
 
        return d.writeNode(ctx, w, node)
```

I would argue **at least `node.Path` must make it into the fatal error message** that `restic dump` croaks with when this problem is encountered.

Curiously, GNU `tar` will happily pack the inode with the "bloated" extended attribute into both its native tar format archive, as well as what GNU tar supposes to be `pax` compatible (via `tar cf /tmp/tardemo.tar /tmp/xfstest_mount/ -H pax --xattrs`), and is also able to fully restore it with all xattr-metadata in both cases, while `pax` proper will contain that the resulting pax-compatible tar format violates the spec (`pax: Extended header record length 59069 is out of range` etc.) when dealing with it (but still re-create the affected files without xattrs). Maybe this warrants a later revisit of `archive/tar` and how it itself stores extended attributes, too...

**Comment by jtru** [comment]
So while the silence problem is fixed (thanks very much for the clean patch!), what are your sentiments towards somehow enabling people to create -as-complete-as-possible tar dumps of their data despite any contained metadata blazing past pax-imposed size limits?

Do you agree that a user-set policy is a worthwhile idea? Or would you rather keep (and document) the current behavior? Or adapt `restic dump` to change the current behavior to another (hard-coded) alternative?

**Comment by greatroar** [comment]
I don't make the decisions, but I think it would be sane to produce a tarball with missing entries and signal the fact that it's incomplete by restic dump's exit status. I've tried implementing that, but found that writing too large a header gets the tar.Writer stuck in an error state. So what the Dumper would need to do is:

1. heuristically check whether the attribute is too large (by a rule that produces no false negatives),
2. if it is, write it to a fresh tar.Writer to be sure,
3. if that does succeed, try again on the main tar.Writer,
4. otherwise, signal the error on a side channel,
5. continue with the next file.

This isn't hard, but it's kind of tricky to get all the paths right and properly tested. The necessary plumbing could be reused to handle some non-fatal errors in archive/zip, especially ones concerning filenames that the zip format cannot represent.

I'm not volunteering to do the work, at least not in the coming months.

**Comment by greatroar** [comment]
Instead of a side channel, we could also use internal/errors.Fatal inside internal/dump, but that would also be a pretty large change.

**Comment by MichaelEischer** [comment]
> This isn't hard, but it's kind of tricky to get all the paths right and properly tested.

I'd be happier if there were a less complex solution for the problem. But having certain files cause the `dump` command to fail seems worse.

**Comment by jtru** [comment]
So the general sentiment is that metadata/xattr preservation behavior in `dump` should follow a "best effort"-approach (in case the target format allows to include any metadata) at all, correct?

**Comment by MichaelEischer** [comment]
Yes, the behavior of `restore` is probably a good guidance. That command for example also continues when some files cannot be written (won't happen for `dump`, but the philosophy still applies).

**Comment by fzlzjerry** [comment]
I'd like to take this on. My plan is to keep the non-zero result when a tar
header cannot be encoded, but detect `tar.ErrFieldTooLong` before it poisons
the output writer, skip that entry, and continue producing a valid archive for
the remaining entries. Other repository and output errors will remain
fail-fast.

I'll add regression coverage with valid entries before and after an entry with
oversized extended attributes, and verify that the returned error still names
the skipped path.
